### PR TITLE
MINIFICPP-558: initial provisioning for CoAP

### PR DESCRIPTION
**Pull request by phrocker**
test framework still under dev

MINIFICPP-683: Remove extraneous std::move

MINIFICPP-558: Add some test capabilities

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by phrocker** [comment]
Since this is pretty large I'm doing a self review of this now.

File: extensions/coap/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+add_library(nanofi-coap-c STATIC ${CSOURCES})
+add_library(minifi-coap STATIC ${SOURCES})
+set_property(TARGET minifi-coap PROPERTY POSITION_INDEPENDENT_CODE ON)
+
+if(CMAKE_THREAD_LIBS_INIT)
+  target_link_libraries(minifi-coap "${CMAKE_THREAD_LIBS_INIT}")
+endif()
+
+  set(BASE_DIR "${CMAKE_CURRENT_BINARY_DIR}/extensions/coap")
+  # determine version of GNUTLSs
```
Comment (phrocker) [inline extensions/coap/CMakeLists.txt]: this is meant as a bit of a todo.

File: extensions/coap/COAPLoader.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  virtual std::string getClassName() override{
+    return "COAPObjectFactory";
+  }
+  /**
+   * Gets the class name for the object
+   * @return class name for the processor.
+   */
+  virtual std::vector<std::string> getClassNames() override{
+    std::vector<std::string> class_names;
+    class_names.push_back("CoapProtocol");
```
Comment (phrocker) [inline extensions/coap/COAPLoader.h]: May want to make CoapServer accessible.

File: extensions/coap/controllerservice/CoapConnector.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+core::Property CoapConnectorService::RemoteServer(core::PropertyBuilder::createProperty("Remote Server")->withDescription("Remote CoAP server")->isRequired(true)->build());
+core::Property CoapConnectorService::Port(core::PropertyBuilder::createProperty("Remote Port")->withDescription("Remote CoAP server port")->isRequired(true)->build());
+core::Property CoapConnectorService::MaxQueueSize(core::PropertyBuilder::createProperty("Max Queue Size")->withDescription("Max queue size for received data ")->isRequired(true)->build());
+
+void CoapConnectorService::initialize() {
+  if (initialized_)
+    return;
+
+  CoapMessaging::getInstance();
```
Comment (phrocker) [inline extensions/coap/controllerservice/CoapConnector.cpp]: Mixed casing of CoAP and coap is quite confusing and hard to follow.

**Comment by apiri** [comment]
reviewing

File: extensions/coap/controllerservice/CoapConnector.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  std::lock_guard<std::mutex> lock(initialization_mutex_);
+
+  ControllerService::initialize();
+
+  initializeProperties();
+
+  initialized_ = true;
+}
+
+void CoapConnectorService::onEnable() {
```
Comment (apiri) [inline extensions/coap/controllerservice/CoapConnector.cpp]: would it be worthwhile to perform a ping as a validation that the endpoint is available?

File: extensions/coap/nanofi/coap_connection.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+  coap_register_option(pdu->ctx, COAP_OPTION_BLOCK2);
+
+  // set the response handler
+  coap_register_response_handler(pdu->ctx, response_handler);
+
+  pdu->session->max_retransmit = 1;
+  pdu->optlist = NULL;
+
+  // add the URI option to the options lsit
```
Comment (apiri) [inline extensions/coap/nanofi/coap_connection.c]: typo

File: extensions/coap/server/CoapServer.h [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+  GET,
+  POST,
+  PUT,
+  DELETE
+};
+
+/**
+ * CoapQuery is the request message sent by the client
+ */
+class CoAPQuery {
```
Comment (apiri) [inline extensions/coap/server/CoapServer.h]: Should probably standardize on the Coap casing used elsewhere

File: extensions/coap/server/CoapServer.h [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+  }
+
+  std::future<uint64_t> future;
+  std::atomic<bool> running_;
+  std::string hostname_;
+  CoapServerContext *server_;
+  static std::map<coap_resource_t*, std::function<CoAPResponse(CoAPQuery)>> functions_;
+  std::map<std::string, CoapEndpoint*> endpoints_;
+//std::vector<CoAPEndpoint*> endpoints_;
```
Comment (apiri) [inline extensions/coap/server/CoapServer.h]: can nix this

File: libminifi/src/c2/C2Agent.cpp [hunk 8]
```diff
@@ -5,7 +5,7 @@
  * this work for additional information regarding copyright ownership.
  * The ASF licenses this file to You under the Apache License, Version 2.0
  * (the "License"); you may not use this file except in compliance with
- * the License.  You may obtain a copy of the License at
+ 8 * the License.  You may obtain a copy of the License at
```
Comment (apiri) [inline libminifi/src/c2/C2Agent.cpp]: crazy 8

File: extensions/coap/controllerservice/CoapConnector.cpp [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  std::string port_str;
+  if (getProperty(RemoteServer.getName(), host_) && !host_.empty() && getProperty(Port.getName(), port_str) && !port_str.empty()) {
```
Comment (apiri) [inline extensions/coap/controllerservice/CoapConnector.cpp]: should get this configuration/usage exposed in our docs

File: extensions/coap/controllerservice/CoapConnector.cpp (hunk 4 above)
Comment (phrocker) [reply extensions/coap/controllerservice/CoapConnector.cpp]: I created this. That's a good point. https://issues.apache.org/jira/browse/MINIFICPP-737 Would you prefer this be part of this PR or a follow up? I don't have a particular preference.

I've updated the PR otherwise.

Comment (apiri) [reply extensions/coap/controllerservice/CoapConnector.cpp]: Certainly fine to do a follow up

**Comment by apiri** [comment]
@phrocker Thanks for the updates.  All looks good here.  I'll take care of merging it in.
